java.lang.NoClassDefFoundError: javax/xml/bind/JAXB

I tried to run the Crowd plugin 2.1.2 on Sonarqube 7.9 on OpenJDK 11 and I run into this problem on startup:

2019.07.02 14:41:31 ERROR web[][o.s.s.p.Platform] Background initialization failed. Stopping SonarQube
java.lang.NoClassDefFoundError: javax/xml/bind/JAXB
    at com.atlassian.crowd.integration.rest.service.RestExecutor.setBody(RestExecutor.java:204)
    at com.atlassian.crowd.integration.rest.service.RestExecutor.post(RestExecutor.java:176)
    at com.atlassian.crowd.integration.rest.service.RestCrowdClient.searchUsers(RestCrowdClient.java:511)
    at com.atlassian.crowd.integration.rest.service.RestCrowdClient.testConnection(RestCrowdClient.java:500)
    at org.sonar.plugins.crowd.CrowdRealm.init(CrowdRealm.java:97)
    at org.sonar.server.user.SecurityRealmFactory.start(SecurityRealmFactory.java:87)
    at org.sonar.core.platform.StartableCloseableSafeLifecyleStrategy.start(StartableCloseableSafeLifecyleStrategy.java:40)
    at org.picocontainer.injectors.AbstractInjectionFactory$LifecycleAdapter.start(AbstractInjectionFactory.java:84)
    at org.picocontainer.behaviors.AbstractBehavior.start(AbstractBehavior.java:169)
    at org.picocontainer.behaviors.Stored$RealComponentLifecycle.start(Stored.java:132)
    at org.picocontainer.behaviors.Stored.start(Stored.java:110)
    at org.picocontainer.DefaultPicoContainer.potentiallyStartAdapter(DefaultPicoContainer.java:1016)
    at org.picocontainer.DefaultPicoContainer.startAdapters(DefaultPicoContainer.java:1009)
    at org.picocontainer.DefaultPicoContainer.start(DefaultPicoContainer.java:767)
    at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:135)
    at org.sonar.server.platform.platformlevel.PlatformLevel.start(PlatformLevel.java:90)
    at org.sonar.server.platform.platformlevel.PlatformLevel4.start(PlatformLevel4.java:545)
    at org.sonar.server.platform.Platform.start(Platform.java:211)
    at org.sonar.server.platform.Platform.startLevel34Containers(Platform.java:185)
    at org.sonar.server.platform.Platform.access$500(Platform.java:46)
    at org.sonar.server.platform.Platform$1.lambda$doRun$0(Platform.java:119)
    at org.sonar.server.platform.Platform$AutoStarterRunnable.runIfNotAborted(Platform.java:371)
    at org.sonar.server.platform.Platform$1.doRun(Platform.java:119)
    at org.sonar.server.platform.Platform$AutoStarterRunnable.run(Platform.java:355)
    at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassNotFoundException: javax.xml.bind.JAXB
    at org.sonar.classloader.ParentFirstStrategy.loadClass(ParentFirstStrategy.java:39)
    at org.sonar.classloader.ClassRealm.loadClass(ClassRealm.java:87)
    at org.sonar.classloader.ClassRealm.loadClass(ClassRealm.java:76)
    ... 25 common frames omitted

I went even deeper... @frdeboffles you need to include 
          <dependency>
            <groupId>com.sun.xml.bind</groupId>
            <artifactId>jaxb-impl</artifactId>
            <version>2.1</version>
          </dependency>

in the sonar-crowd pom.xml file, also download jaxb-impl-2.1 into sonarqube-7.9/lib/common ... make sure that all of the dependencies match in version, otherwise you will get a REFLECTION error. 

Then after all of that troubleshooting, 

```
2019.07.03 20:18:39 INFO  web[][o.s.p.ProcessEntryPoint] Starting web
2019.07.03 20:18:40 INFO  web[][o.a.t.u.n.NioSelectorPool] Using a shared selector for servlet write/read
2019.07.03 20:18:41 ERROR web[][o.s.s.a.EmbeddedTomcat] Fail to start web server
org.apache.catalina.LifecycleException: Failed to start component [StandardServer[-1]]
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:167)
        at org.apache.catalina.startup.Tomcat.start(Tomcat.java:427)
        at org.sonar.server.app.EmbeddedTomcat.start(EmbeddedTomcat.java:65)
        at org.sonar.server.app.WebServer.start(WebServer.java:52)
        at org.sonar.process.ProcessEntryPoint.launch(ProcessEntryPoint.java:97)
        at org.sonar.process.ProcessEntryPoint.launch(ProcessEntryPoint.java:81)
        at org.sonar.server.app.WebServer.main(WebServer.java:99)
Caused by: org.apache.catalina.LifecycleException: Failed to start component [StandardService[Tomcat]]
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:167)
        at org.apache.catalina.core.StandardServer.startInternal(StandardServer.java:793)
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
        ... 6 common frames omitted
Caused by: org.apache.catalina.LifecycleException: Failed to start component [StandardEngine[Tomcat]]
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:167)
        at org.apache.catalina.core.StandardService.startInternal(StandardService.java:422)
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
        ... 8 common frames omitted
Caused by: org.apache.catalina.LifecycleException: A child container failed during start
        at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:953)
        at org.apache.catalina.core.StandardEngine.startInternal(StandardEngine.java:262)
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
        ... 10 common frames omitted
Caused by: java.util.concurrent.ExecutionException: org.apache.catalina.LifecycleException: Failed to start component [StandardEngine[Tomcat].StandardHost[localhost]]
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:942)
        ... 12 common frames omitted
Caused by: org.apache.catalina.LifecycleException: Failed to start component [StandardEngine[Tomcat].StandardHost[localhost]]
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:167)
        at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1423)
        at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1413)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.apache.catalina.LifecycleException: A child container failed during start
        at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:953)
        at org.apache.catalina.core.StandardHost.startInternal(StandardHost.java:872)
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
        ... 6 common frames omitted
Caused by: org.apache.tomcat.util.MultiThrowable: null
        at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:946)
        ... 8 common frames omitted
```

I don't want to give up but, I think with all this hacked up stuff. i'm kind of out of options.

Here was the error after fixing the jaxb-api / jaxb-impl / jaxb-core issue:

`java.lang.NoClassDefFoundError: com/sun/istack/Pool
        at com.sun.xml.bind.v2.runtime.JAXBContextImpl$JAXBContextBuilder.build(JAXBContextImpl.java:1156)
        at com.sun.xml.bind.v2.ContextFactory.createContext(ContextFactory.java:165)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at javax.xml.bind.ContextFinder.newInstance(ContextFinder.java:297)
        at javax.xml.bind.ContextFinder.newInstance(ContextFinder.java:286)
        at javax.xml.bind.ContextFinder.find(ContextFinder.java:409)
        at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:721)
        at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:662)
        at javax.xml.bind.JAXB$Cache.<init>(JAXB.java:127)
        at javax.xml.bind.JAXB.getContext(JAXB.java:154)
        at javax.xml.bind.JAXB._marshal(JAXB.java:578)
        at javax.xml.bind.JAXB.marshal(JAXB.java:447)
        at com.atlassian.crowd.integration.rest.service.RestExecutor.setBody(RestExecutor.java:204)
        at com.atlassian.crowd.integration.rest.service.RestExecutor.post(RestExecutor.java:176)
        at com.atlassian.crowd.integration.rest.service.RestCrowdClient.searchUsers(RestCrowdClient.java:511)
        at com.atlassian.crowd.integration.rest.service.RestCrowdClient.testConnection(RestCrowdClient.java:500)
        at org.sonar.plugins.crowd.CrowdRealm.init(CrowdRealm.java:97)
        at org.sonar.server.user.SecurityRealmFactory.start(SecurityRealmFactory.java:87)
        at org.sonar.core.platform.StartableCloseableSafeLifecyleStrategy.start(StartableCloseableSafeLifecyleStrategy.java:40)
        at org.picocontainer.injectors.AbstractInjectionFactory$LifecycleAdapter.start(AbstractInjectionFactory.java:84)
        at org.picocontainer.behaviors.AbstractBehavior.start(AbstractBehavior.java:169)
        at org.picocontainer.behaviors.Stored$RealComponentLifecycle.start(Stored.java:132)
        at org.picocontainer.behaviors.Stored.start(Stored.java:110)
        at org.picocontainer.DefaultPicoContainer.potentiallyStartAdapter(DefaultPicoContainer.java:1016)
        at org.picocontainer.DefaultPicoContainer.startAdapters(DefaultPicoContainer.java:1009)
        at org.picocontainer.DefaultPicoContainer.start(DefaultPicoContainer.java:767)
        at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:135)
        at org.sonar.server.platform.platformlevel.PlatformLevel.start(PlatformLevel.java:90)
        at org.sonar.server.platform.platformlevel.PlatformLevel4.start(PlatformLevel4.java:545)
        at org.sonar.server.platform.Platform.start(Platform.java:211)
        at org.sonar.server.platform.Platform.startLevel34Containers(Platform.java:185)
        at org.sonar.server.platform.Platform.access$500(Platform.java:46)
        at org.sonar.server.platform.Platform$1.lambda$doRun$0(Platform.java:119)
        at org.sonar.server.platform.Platform$AutoStarterRunnable.runIfNotAborted(Platform.java:371)
        at org.sonar.server.platform.Platform$1.doRun(Platform.java:119)
        at org.sonar.server.platform.Platform$AutoStarterRunnable.run(Platform.java:355)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassNotFoundException: com.sun.istack.Pool
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
        ... 40 common frames omitted
`

--- a/pom.xml
+++ b/pom.xml
@@ -56,6 +56,7 @@
     </issueManagement>
 
     <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <sonar.version>6.7</sonar.version>
         <sonar.pluginName>Crowd</sonar.pluginName>
         <sonar.pluginClass>org.sonar.plugins.crowd.CrowdPlugin</sonar.pluginClass>
@@ -91,6 +92,26 @@
             <artifactId>jcl-over-slf4j</artifactId>
             <version>1.7.25</version>
         </dependency>
+        <dependency>
+            <groupId>javax.xml.bind</groupId>
+            <artifactId>jaxb-api</artifactId>
+            <version>2.3.0</version>
+        </dependency>
+        <dependency>
+            <groupId>javax.activation</groupId>
+            <artifactId>activation</artifactId>
+            <version>1.1</version>
+        </dependency>
+        <dependency>
+            <groupId>org.glassfish.jaxb</groupId>
+            <artifactId>jaxb-runtime</artifactId>
+            <version>2.3.0</version>
+        </dependency>
+        <dependency>
+            <groupId>com.sun.xml.bind</groupId>
+            <artifactId>jaxb-core</artifactId>
+            <version>2.3.0</version>
+        </dependency>
         <!-- unit tests -->
         <dependency>
             <groupId>org.sonarsource.sonarqube</groupId>
@@ -118,12 +139,13 @@
             <url>https://packages.atlassian.com/maven-public-legacy-local</url>
         </repository>
     </repositories>
-    
+
     <build>
         <plugins>
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.8.1</version>
                 <configuration>
                     <source>8</source>
                     <target>8</target>

--- a/src/main/java/org/sonar/plugins/crowd/CrowdAuthenticator.java
+++ b/src/main/java/org/sonar/plugins/crowd/CrowdAuthenticator.java
@@ -53,28 +53,38 @@ public class CrowdAuthenticator extends Authenticator {
 
 
   public boolean authenticate(String login, String password) {
+    // Had to add that as from "not really a good idea" in
+    // https://stackoverflow.com/questions/51518781/jaxb-not-available-on-tomcat-9-and-java-9-10
+    ClassLoader threadClassLoader = Thread.currentThread().getContextClassLoader();
     try {
-      client.authenticateUser(login, password);
-      return true;
-    } catch (UserNotFoundException e) {
-      LOG.debug("User {} not found", login);
-      return false;
-    } catch (InactiveAccountException e) {
-      LOG.debug("User {} is not active", login);
-      return false;
-    } catch (ExpiredCredentialException e) {
-      LOG.debug("Credentials of user {} have expired", login);
-      return false;
-    } catch (ApplicationPermissionException e) {
-      LOG.error("The application is not permitted to perform the requested operation"
-        + " on the crowd server", e);
-      return false;
-    } catch (InvalidAuthenticationException e) {
-      LOG.debug("Invalid credentials for user {}", login);
-      return false;
-    } catch (OperationFailedException e) {
-      LOG.error("Unable to authenticate user " + login, e);
-      return false;
+      // This will enforce the crowClient to use the plugin classloader
+      Thread.currentThread().setContextClassLoader(this.getClass().getClassLoader());
+      try {
+        client.authenticateUser(login, password);
+        return true;
+      } catch (UserNotFoundException e) {
+        LOG.debug("User {} not found", login);
+        return false;
+      } catch (InactiveAccountException e) {
+        LOG.debug("User {} is not active", login);
+        return false;
+      } catch (ExpiredCredentialException e) {
+        LOG.debug("Credentials of user {} have expired", login);
+        return false;
+      } catch (ApplicationPermissionException e) {
+        LOG.error("The application is not permitted to perform the requested operation"
+          + " on the crowd server", e);
+        return false;
+      } catch (InvalidAuthenticationException e) {
+        LOG.debug("Invalid credentials for user {}", login);
+        return false;
+      } catch (OperationFailedException e) {
+        LOG.error("Unable to authenticate user " + login, e);
+        return false;
+      }
+    } finally {
+      // Bring back the original class loader for the thread
+      Thread.currentThread().setContextClassLoader(threadClassLoader);
     }
   }
 }

--- a/src/main/java/org/sonar/plugins/crowd/CrowdGroupsProvider.java
+++ b/src/main/java/org/sonar/plugins/crowd/CrowdGroupsProvider.java
@@ -60,7 +60,17 @@ public class CrowdGroupsProvider extends ExternalGroupsProvider {
   private Collection<String> getGroupsForUser(String username, int start, int pageSize)
     throws UserNotFoundException, OperationFailedException, InvalidAuthenticationException,
     ApplicationPermissionException {
-    return transform(crowdClient.getGroupsForNestedUser(username, start, pageSize), GROUP_TO_STRING);
+    // Had to add that as from "not really a good idea" in
+    // https://stackoverflow.com/questions/51518781/jaxb-not-available-on-tomcat-9-and-java-9-10
+    ClassLoader threadClassLoader = Thread.currentThread().getContextClassLoader();
+    try {
+      // This will enforce the crowClient to use the plugin classloader
+      Thread.currentThread().setContextClassLoader(this.getClass().getClassLoader());
+      return transform(crowdClient.getGroupsForNestedUser(username, start, pageSize), GROUP_TO_STRING);
+    } finally {
+      // Bring back the original class loader for the thread
+      Thread.currentThread().setContextClassLoader(threadClassLoader);
+    }
   }
 
   private List<String> getGroupsForUser(String username)

--- a/src/main/java/org/sonar/plugins/crowd/CrowdRealm.java
+++ b/src/main/java/org/sonar/plugins/crowd/CrowdRealm.java
@@ -93,16 +93,26 @@ public class CrowdRealm extends SecurityRealm {
     this.authenticator = new CrowdAuthenticator(crowdClient);
     this.usersProvider = new CrowdUsersProvider(crowdClient);
     this.groupsProvider = new CrowdGroupsProvider(crowdClient);
+    // Had to add that as from "not really a good idea" in
+    // https://stackoverflow.com/questions/51518781/jaxb-not-available-on-tomcat-9-and-java-9-10
+    ClassLoader threadClassLoader = Thread.currentThread().getContextClassLoader();
     try {
-      crowdClient.testConnection();
-      LOG.info("Crowd configuration is valid, connection test successful.");
-    } catch (OperationFailedException e) {
-      throw new SonarException("Unable to test connection to crowd", e);
-    } catch (InvalidAuthenticationException e) {
-      throw new SonarException("Application name and password are incorrect", e);
-    } catch (ApplicationPermissionException e) {
-      throw new SonarException("The application is not permitted to perform the requested "
-        + "operation on the crowd server", e);
+      // This will enforce the crowClient to use the plugin classloader
+      Thread.currentThread().setContextClassLoader(this.getClass().getClassLoader());
+      try {
+        crowdClient.testConnection();
+        LOG.info("Crowd configuration is valid, connection test successful.");
+      } catch (OperationFailedException e) {
+        throw new SonarException("Unable to test connection to crowd", e);
+      } catch (InvalidAuthenticationException e) {
+        throw new SonarException("Application name and password are incorrect", e);
+      } catch (ApplicationPermissionException e) {
+        throw new SonarException("The application is not permitted to perform the requested "
+            + "operation on the crowd server", e);
+      }
+    } finally {
+      // Bring back the original class loader for the thread
+      Thread.currentThread().setContextClassLoader(threadClassLoader);
     }
   }
 

--- a/src/main/java/org/sonar/plugins/crowd/CrowdUsersProvider.java
+++ b/src/main/java/org/sonar/plugins/crowd/CrowdUsersProvider.java
@@ -47,30 +47,39 @@ public class CrowdUsersProvider extends ExternalUsersProvider {
   @Override
   public UserDetails doGetUserDetails(String username) {
     LOG.debug("Looking up user details for user {}", username);
-
+    // Had to add that as from "not really a good idea" in
+    // https://stackoverflow.com/questions/51518781/jaxb-not-available-on-tomcat-9-and-java-9-10
+    ClassLoader threadClassLoader = Thread.currentThread().getContextClassLoader();
     try {
-      User user = crowdClient.getUser(username);
-      UserDetails details = new UserDetails();
-      if (user.getDisplayName() != null) {
-        details.setName(user.getDisplayName());
+      // This will enforce the crowClient to use the plugin classloader
+      Thread.currentThread().setContextClassLoader(this.getClass().getClassLoader());
+      try {
+        User user = crowdClient.getUser(username);
+        UserDetails details = new UserDetails();
+        if (user.getDisplayName() != null) {
+          details.setName(user.getDisplayName());
+        }
+        if (user.getEmailAddress() != null) {
+          details.setEmail(user.getEmailAddress());
+        }
+        return details;
+      } catch (UserNotFoundException e) {
+        return null; // API contract for ExternalUsersProvider
+      } catch (OperationFailedException e) {
+        throw new SonarException("Unable to retrieve user details for user" + username
+          + " from crowd.", e);
+      } catch (ApplicationPermissionException e) {
+        throw new SonarException(
+          "Unable to retrieve user details for user" + username
+            + " from crowd. The application is not permitted to perform the "
+            + "requested operation on the crowd server.", e);
+      } catch (InvalidAuthenticationException e) {
+        throw new SonarException("Unable to retrieve user details for user" + username
+          + " from crowd. The application name and password are incorrect.", e);
       }
-      if (user.getEmailAddress() != null) {
-        details.setEmail(user.getEmailAddress());
-      }
-      return details;
-    } catch (UserNotFoundException e) {
-      return null; // API contract for ExternalUsersProvider
-    } catch (OperationFailedException e) {
-      throw new SonarException("Unable to retrieve user details for user" + username
-        + " from crowd.", e);
-    } catch (ApplicationPermissionException e) {
-      throw new SonarException(
-        "Unable to retrieve user details for user" + username
-          + " from crowd. The application is not permitted to perform the "
-          + "requested operation on the crowd server.", e);
-    } catch (InvalidAuthenticationException e) {
-      throw new SonarException("Unable to retrieve user details for user" + username
-        + " from crowd. The application name and password are incorrect.", e);
+    } finally {
+      // Bring back the original class loader for the thread
+      Thread.currentThread().setContextClassLoader(threadClassLoader);
     }
   }
 

